File: backend/app/shared/schemas/patient.py
"""Patient Pydantic schemas."""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, EmailStr, Field, ConfigDict


# ===== Patient Schemas =====

class PatientBase(BaseModel):
    """Base patient fields."""

    full_name: str = Field(..., min_length=1, max_length=255)
    date_of_birth: date | None = None
    gender: Literal["male", "female", "other"] | None = None
    blood_group: Literal["A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"] | None = None

    # Contact
    phone: str | None = Field(None, max_length=20)
    email: EmailStr | None = None
    whatsapp: str | None = Field(None, max_length=20)

    # Address
    address: str | None = None
    division_id: int | None = None
    district_id: int | None = None
    upazila_id: int | None = None

    # Medical
    chief_complaint: str | None = None
    medical_history: str | None = None

    # Photo
    photo_url: str | None = Field(None, max_length=500)

    # Follow-up
    next_visit_date: date | None = None


class PatientCreate(PatientBase):
    """Create patient request."""
    pass


class PatientUpdate(BaseModel):
    """Update patient request (all fields optional)."""

    full_name: str | None = Field(None, min_length=1, max_length=255)
    date_of_birth: date | None = None
    gender: Literal["male", "female", "other"] | None = None
    blood_group: Literal["A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"] | None = None

    phone: str | None = Field(None, max_length=20)
    email: EmailStr | None = None
    whatsapp: str | None = Field(None, max_length=20)

    address: str | None = None
    division_id: int | None = None
    district_id: int | None = None
    upazila_id: int | None = None

    chief_complaint: str | None = None
    medical_history: str | None = None
    photo_url: str | None = Field(None, max_length=500)
    next_visit_date: date | None = None
    is_active: bool | None = None


class PatientResponse(PatientBase):
    """Patient response with all fields."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    is_active: bool
    created_at: datetime
    updated_at: datetime | None
    created_by: str | None
    updated_by: str | None


class PatientListItem(BaseModel):
    """Lightweight patient list item."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    full_name: str
    phone: str | None
    date_of_birth: date | None
    gender: str | None
    next_visit_date: date | None
    is_active: bool


class PatientSearchResult(PatientListItem):
    """Patient search result with additional context."""

    model_config = ConfigDict(from_attributes=True)

    email: str | None
    chief_complaint: str | None
    created_at: datetime


# ===== Patient Tag Schemas =====

class PatientTagBase(BaseModel):
    """Base patient tag fields."""

    tag_type: Literal["special_case", "chronic", "treatment", "allergy"]
    tag_value: str = Field(..., min_length=1, max_length=255)
    notes: str | None = None


class PatientTagCreate(PatientTagBase):
    """Create patient tag request."""
    patient_id: str | None = None  # Optional - will be set from URL path in route handler


class PatientTagUpdate(BaseModel):
    """Update patient tag request."""

    tag_type: Literal["special_case", "chronic", "treatment", "allergy"] | None = None
    tag_value: str | None = Field(None, min_length=1, max_length=255)
    notes: str | None = None


class PatientTagResponse(PatientTagBase):
    """Patient tag response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    patient_id: str
    tenant_id: str
    created_at: datetime
    updated_at: datetime | None


# ===== Patient Diagnosis Schemas =====

class PatientDiagnosisBase(BaseModel):
    """Base patient diagnosis fields."""

    description: str = Field(..., min_length=1)
    icd_code: str | None = Field(None, max_length=20)
    diagnosed_at: date


class PatientDiagnosisCreate(PatientDiagnosisBase):
    """Create patient diagnosis request."""

    patient_id: str | None = None  # Optional - will be set from URL path in route handler
    visit_id: str | None = None


class PatientDiagnosisUpdate(BaseModel):
    """Update patient diagnosis request."""

    description: str | None = Field(None, min_length=1)
    icd_code: str | None = Field(None, max_length=20)
    diagnosed_at: date | None = None
    is_active: bool | None = None


class PatientDiagnosisResponse(PatientDiagnosisBase):
    """Patient diagnosis response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    patient_id: str
    visit_id: str | None
    tenant_id: str
    is_active: bool
    created_at: datetime
    updated_at: datetime | None

File: frontend/src/types/patient.ts
// PatientListItem — lightweight list view (matches backend PatientListItem)
export interface PatientListItem {
  id: string
  full_name: string
  phone: string | null
  date_of_birth: string | null
  gender: 'male' | 'female' | 'other' | null
  next_visit_date: string | null
  is_active: boolean
}

// Patient — full record (matches backend PatientResponse)
export interface Patient {
  id: string
  tenant_id: string
  full_name: string
  date_of_birth: string | null
  gender: 'male' | 'female' | 'other' | null
  blood_group: 'A+' | 'A-' | 'B+' | 'B-' | 'AB+' | 'AB-' | 'O+' | 'O-' | null
  phone: string | null
  email: string | null
  whatsapp: string | null
  address: string | null
  division_id: number | null
  district_id: number | null
  upazila_id: number | null
  chief_complaint: string | null
  medical_history: string | null
  photo_url: string | null
  next_visit_date: string | null
  is_active: boolean
  created_at: string
  updated_at: string | null
  created_by: string | null
  updated_by: string | null
}

// PatientCreateRequest — matches backend PatientCreate
export interface PatientCreateRequest {
  full_name: string
  date_of_birth?: string
  gender?: 'male' | 'female' | 'other'
  blood_group?: 'A+' | 'A-' | 'B+' | 'B-' | 'AB+' | 'AB-' | 'O+' | 'O-'
  phone?: string
  email?: string
  whatsapp?: string
  address?: string
  division_id?: number
  district_id?: number
  upazila_id?: number
  chief_complaint?: string
  medical_history?: string
  photo_url?: string
  next_visit_date?: string
}

// PatientUpdateRequest — all fields optional, plus is_active toggle
export type PatientUpdateRequest = Partial<PatientCreateRequest> & {
  is_active?: boolean
}

export interface PatientListParams {
  limit?: number
  offset?: number
  search?: string
  is_active?: boolean
}

export interface PatientTag {
  id: number
  tenant_id: string
  patient_id: string
  tag_type: 'special_case' | 'chronic' | 'treatment' | 'allergy'
  tag_value: string
  notes: string | null
  created_at: string
  updated_at: string | null
}

export interface PatientDiagnosis {
  id: number
  tenant_id: string
  patient_id: string
  visit_id: string | null
  description: string
  icd_code: string | null
  diagnosed_at: string
  is_active: boolean
  created_at: string
  updated_at: string | null
}

// Geographic types
export interface Division {
  id: number
  name_en: string
  name_bn: string
  code: string
}

export interface District {
  id: number
  division_id: number
  name_en: string
  name_bn: string
  code: string
}

export interface Upazila {
  id: number
  district_id: number
  name_en: string
  name_bn: string
  code: string
}
